### ISSUE TITLE
flush entities in an transaction and rollback when during error

### ISSUE THREAD

**Issue by mamkkl** [doctrine/doctrine2#5759]
in the flush procedure of entity manager, it seems changes will be committed at once when calling the flush function even I use the beginTransaction(). I am using doctrine/orm@dev-master
```
$conn = $em->getConnection();
$conn->beginTransaction();
try{
   // $conn to delete existing data
  // create $entity1
   $em->persist($entity1);
   $em->flush();
   try{
      // create $entity2
      $em->persist($entity2);
      // error in here and rollback to the beginning before taking any actions, 
      //however, it's not rolling back to the beginning point.
      $em->flush(); 
      $em->getConnection()->commit();
   }catch(){
      $em->getConnection()->rollBack();
   }
   $em->getConnection()->commit();
}catch(){
   $em->getConnection()->rollBack();
}
```

**Comment by Ocramius** [comment]
@mamkkl could you log the queries being executed, and when? What sequence of queries did you expect?

**Comment by DHager** [comment]
@mamkkl I believe that's how it is supposed  to work. `flush()` sends all the `INSERT`/`UPDATE`/`DELETE` items out to the database, and then `commit()` sends `COMMIT` as a final message to tell the database to commit everything it's already seen.

**Comment by mamkkl** [comment]
Yes, what I am thinking is, once it reach an error and rollback in catch, the whole process should be rollback, even some `flush()` has been reached.

**Comment by Jean85** [comment]
@mamkkl are you checking the changes at the DB level or inside your entities?

**Comment by mamkkl** [comment]
I am checking changes in DB level and found entities those were `flush()` successfully can't be rollback (records exist in DB).

**Comment by mamkkl** [comment]
@DHager So, I can't do `CUD` in a transaction via entities?

**Comment by chinmayshah24** [comment]
@mamkkl how were you able to get around this? I see this too

**Comment by Ocramius** [comment]
@mamkkl @chinmayshah24 maybe you are using a non-transactional storage engine? It is well known that [`MyISAM` does not support transactions](https://stackoverflow.com/questions/3404467/mysql-rollback-on-the-myisam-engine).

**Comment by chinmayshah24** [comment]
Oh yes. Legacy tables, didn't check that. Thanks @Ocramius 
I can confirm this works.
